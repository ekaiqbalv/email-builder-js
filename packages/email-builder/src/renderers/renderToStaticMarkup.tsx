import React from 'react';
import { renderToStaticMarkup as baseRenderToStaticMarkup } from 'react-dom/server';

import Reader, { TReaderDocument } from '../Reader/core';
import {
  collectEmailStyles,
  extractEmailClassNames,
  mergeEmailStyles,
  organizeEmailStylesByMedia,
} from '../utils/email-style';

type TOptions = {
  rootBlockId: string;
};
export default function renderToStaticMarkup(document: TReaderDocument, { rootBlockId }: TOptions) {
  const styleElements = window.document.querySelectorAll('[data-emotion]');
  const bodyContent = baseRenderToStaticMarkup(<Reader document={document} rootBlockId={rootBlockId} />);

  const uniqueClassNames = extractEmailClassNames(bodyContent);
  const styles = collectEmailStyles(styleElements, uniqueClassNames);
  const styleGroups = organizeEmailStylesByMedia(styles);
  const combinedStyles = mergeEmailStyles(styleGroups);

  return `
    <!DOCTYPE html>
    <html>
      <head>
        <style>${combinedStyles}</style>
      </head>
      <body>${bodyContent}</body>
    </html>
  `;
}
