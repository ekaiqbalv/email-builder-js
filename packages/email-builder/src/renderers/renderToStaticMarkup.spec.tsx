/**
 * @jest-environment node
 */

import renderToStaticMarkup from './renderToStaticMarkup';

describe('renderToStaticMarkup', () => {
  beforeEach(() => {
    global.window = {
      document: {
        querySelectorAll: jest.fn().mockReturnValue([]),
      },
    } as any;
  });

  it('renders into a string', () => {
    const result = renderToStaticMarkup(
      {
        root: {
          type: 'Container',
          data: {
            props: {
              childrenIds: [],
            },
          },
        },
      },
      { rootBlockId: 'root' }
    );
    expect(result).toMatchSnapshot();
  });
});
