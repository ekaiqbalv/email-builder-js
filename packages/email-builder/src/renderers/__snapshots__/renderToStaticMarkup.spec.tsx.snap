// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renderToStaticMarkup renders into a string 1`] = `
"
    <!DOCTYPE html>
    <html>
      <head>
        <style></style>
      </head>
      <body><div></div></body>
    </html>
  "
`;
